pkg/model/mselection/heaviest.go

```diff
@@ -20,137 +19,230 @@ import (
-func (s *HeaviestSelector) ResetCone(tipHash hornet.Hash) error {
-	s.Lock()
-	defer s.Unlock()
+// applyTip set all bits of all referenced transactions of the tip in all existing tips to zero.
+// this way we can track which parts of the cone would already be referenced by this tip, and
+// correctly calculate the weight of the remaining tips.
+func (il *itemList) applyTip(tip *selectedTip) {
+
+	il.removeTip(tip)
 
-	choosenTip, exists := s.approvers[string(tipHash)]
-	if !exists {
-		return ErrWrongReference
+	// set all bits of all referenced transactions in all existing tips to zero
+	for _, otherTip := range il.items {
+		otherTip.refs.InPlaceDifference(tip.item.refs)
 	}
+}
 
-	// remove the used tip from the tips list
-	s.removeTip(choosenTip)
+// removeTip removes the tip from the list if items.
+func (il *itemList) removeTip(tip *selectedTip) {
+	il.items[tip.index] = il.items[len(il.items)-1]
+	il.items[len(il.items)-1] = nil
+	il.items = il.items[:len(il.items)-1]
+}
 
-	// set all bits of all referenced transactions in all existing tips to zero
-	for e := s.tips.Front(); e != nil; e = e.Next() {
-		e.Value.(*item).refs.InPlaceDifference(choosenTip.refs)
+// New creates a new HeaviestSelector instance.
+func New(minHeaviestBranchUnconfirmedTransactionsThreshold int, maxHeaviestBranchTipsPerCheckpoint int, randomTipsPerCheckpoint int) *HeaviestSelector {
+	s := &HeaviestSelector{
+		minHeaviestBranchUnconfirmedTransactionsThreshold: minHeaviestBranchUnconfirmedTransactionsThreshold,
+		maxHeaviestBranchTipsPerCheckpoint:                maxHeaviestBranchTipsPerCheckpoint,
+		randomTipsPerCheckpoint:                           randomTipsPerCheckpoint,
 	}
+	s.reset()
+	return s
+}
+
+// reset resets the approvers and tips list of s.
+func (s *HeaviestSelector) reset() {
+	s.Lock()
+	defer s.Unlock()
 
-	return nil
+	// create an empty map
+	s.approvers = make(map[trinary.Hash]*item)
+
+	// create an empty list
+	s.tips = list.New()
 }
 
-// selectTip selects a tip to be used for the next milestone.
-// It returns a tip, confirming the most transactions in the future cone.
-// The selection can be cancelled anytime via the provided context. In this case, it returns the current best solution.
-// selectTip be called concurrently with other HeaviestSelector methods. However, it only considers the tips
-// that were present at the beginning of the selectTip call.
-// TODO: add a proper interface for ms selection that is used by the coordinator
-func (s *HeaviestSelector) selectTip(ctx context.Context) (hornet.Hash, error) {
-	// copy the tips to release the lock at allow faster iteration
-	tips := s.tipItems()
+// selectTip selects a tip to be used for the next checkpoint.
+// it returns a tip, confirming the most transactions in the future cone.
+// the selection can be cancelled anytime via the provided context. in this case, it returns the current best solution.
```

Comment:
Could you add in the comment what the returned 'uint' represents? It is actually used to determine threshold conditions, so good to get documented.